File: workspace/platf-bizz-prjs/drr-tnt-swfabrik-europe/food-market-06-service/src/Adapter/In/Rest/FoodTradingController.cs
using System.Text.Json;
using Darueira.FoodMarket.Service06.Adapter.In.Rest.Dtos;
using Darueira.FoodMarket.Service06.Domain.Commands;
using Darueira.FoodMarket.Service06.Port.In;
using Microsoft.AspNetCore.Mvc;

namespace Darueira.FoodMarket.Service06.Adapter.In.Rest;

[ApiController]
[Route("api/food-tradings")]
public class FoodTradingController : ControllerBase
{
    private readonly IFoodTradingUseCase _useCase;

    public FoodTradingController(IFoodTradingUseCase useCase)
    {
        _useCase = useCase;
    }

    [HttpPost]
    [ProducesResponseType(typeof(FoodTradingResponse), StatusCodes.Status201Created)]
    public async Task<IActionResult> CreateTrading([FromBody] CreateTradingRequest request, CancellationToken ct)
    {
        try
        {
            var command = new CreateFoodTradingCommand
            {
                ItemName = request.ItemName,
                Quantity = request.Quantity,
                UnitPrice = request.UnitPrice,
                TraderName = request.TraderName
            };

            var created = await _useCase.CreateTradingAsync(command, ct);
            return StatusCode(StatusCodes.Status201Created, FoodTradingResponse.FromDomain(created));
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    [HttpGet]
    [ProducesResponseType(typeof(IReadOnlyList<FoodTradingResponse>), StatusCodes.Status200OK)]
    public async Task<IActionResult> ListTradings(CancellationToken ct)
    {
        var tradings = await _useCase.ListTradingsAsync(ct);
        var response = tradings.Select(FoodTradingResponse.FromDomain).ToList();
        return Ok(response);
    }

    [HttpGet("stream")]
    public async Task StreamTradings(CancellationToken ct)
    {
        Response.Headers.Append("Content-Type", "text/event-stream");
        Response.Headers.Append("Cache-Control", "no-cache");
        Response.Headers.Append("Connection", "keep-alive");

        // Send initial event
        await Response.WriteAsync("event: INIT\ndata: {\"message\":\"Connected to Food Trading Live SSE Stream (Service 06 - .NET 8 / C#)\"}\n\n", ct);
        await Response.Body.FlushAsync(ct);

        await foreach (var trading in _useCase.SubscribeStreamAsync(ct))
        {
            var json = JsonSerializer.Serialize(FoodTradingResponse.FromDomain(trading));
            await Response.WriteAsync($"event: FOOD_TRADING_EVENT\ndata: {json}\n\n", ct);
            await Response.Body.FlushAsync(ct);
        }
    }
}

File: apps/examples/tenant-xpto/food-market-05-service/src/adapter/in/rest/food-trading.controller.ts
import { Body, Controller, Get, Inject, Post, Sse } from '@nestjs/common';
import { ApiOperation, ApiResponse, ApiTags } from '@nestjs/swagger';
import { Observable } from 'rxjs';
import { CreateTradingDto } from './dto/create-trading.dto';
import { FoodTradingDto } from './dto/food-trading.dto';
import { FOOD_TRADING_USE_CASE, FoodTradingUseCasePort } from '../../../port/in/food-trading-use-case.port';
import { CreateFoodTradingCommand } from '../../../domain/create-food-trading.command';

@ApiTags('Food Tradings (NestJS)')
@Controller('api/food-tradings')
export class FoodTradingController {
  constructor(
    @Inject(FOOD_TRADING_USE_CASE)
    private readonly useCase: FoodTradingUseCasePort,
  ) {}

  @Post()
  @ApiOperation({ summary: 'F01.1: Create Food Trading in NestJS' })
  @ApiResponse({ status: 201, type: FoodTradingDto })
  async createTrading(@Body() dto: CreateTradingDto): Promise<FoodTradingDto> {
    const cmd = new CreateFoodTradingCommand({
      itemName: dto.itemName,
      quantity: dto.quantity,
      unitPrice: dto.unitPrice,
      traderName: dto.traderName,
    });
    const created = await this.useCase.createTrading(cmd);
    return FoodTradingDto.fromDomain(created);
  }

  @Get()
  @ApiOperation({ summary: 'F01.2: List Food Tradings from schema schm05' })
  @ApiResponse({ status: 200, type: [FoodTradingDto] })
  async listTradings(): Promise<FoodTradingDto[]> {
    const list = await this.useCase.listTradings();
    return list.map((item) => FoodTradingDto.fromDomain(item));
  }

  @Sse('stream')
  @ApiOperation({ summary: 'F01.3: Real-Time SSE Stream of Food Trading Events' })
  streamTradings(): Observable<any> {
    return this.useCase.subscribeStream();
  }
}
